src/trio/_core/_ki.py

```diff
@@ -117,90 +125,37 @@ def currently_ki_protected() -> bool:
-) -> Callable[[Callable[ArgsT, RetT]], Callable[ArgsT, RetT]]:
-    # The "ignore[return-value]" below is because the inspect functions cast away the
-    # original return type of fn, making it just CoroutineType[Any, Any, Any] etc.
-    # ignore[misc] is because @wraps() is passed a callable with Any in the return type.
-    def decorator(fn: Callable[ArgsT, RetT]) -> Callable[ArgsT, RetT]:
-        # In some version of Python, isgeneratorfunction returns true for
-        # coroutine functions, so we have to check for coroutine functions
-        # first.
-        if inspect.iscoroutinefunction(fn):
-
-            @wraps(fn)
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:  # type: ignore[misc]
-                # See the comment for regular generators below
-                coro = fn(*args, **kwargs)
-                assert coro.cr_frame is not None, "Coroutine frame should exist"
-                coro.cr_frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return coro  # type: ignore[return-value]
-
-            return wrapper
-        elif inspect.isgeneratorfunction(fn):
-
-            @wraps(fn)
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:  # type: ignore[misc]
-                # It's important that we inject this directly into the
-                # generator's locals, as opposed to setting it here and then
-                # doing 'yield from'. The reason is, if a generator is
-                # throw()n into, then it may magically pop to the top of the
-                # stack. And @contextmanager generators in particular are a
-                # case where we often want KI protection, and which are often
-                # thrown into! See:
-                #     https://bugs.python.org/issue29590
-                gen = fn(*args, **kwargs)
-                gen.gi_frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return gen  # type: ignore[return-value]
-
-            return wrapper
-        elif inspect.isasyncgenfunction(fn) or legacy_isasyncgenfunction(fn):
-
-            @wraps(fn)  # type: ignore[arg-type]
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:  # type: ignore[misc]
-                # See the comment for regular generators above
-                agen = fn(*args, **kwargs)
-                agen.ag_frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return agen  # type: ignore[return-value]
-
-            return wrapper
-        else:
 
-            @wraps(fn)
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:
-                sys._getframe().f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return fn(*args, **kwargs)
+_T_supports_code = TypeVar("_T_supports_code", bound=_SupportsCode)
 
-            return wrapper
 
-    return decorator
+def enable_ki_protection(f: _T_supports_code, /) -> _T_supports_code:
```

Comment:
there's a slight user facing change to how this works, if you do something like:

```python
def foo():
    yield

@enable_ki_protection
def baz():
    return foo()
```

then ki protection is only enabled on baz, and not on foo, wheras before ki protection would only be on foo but not on baz